File: java/sum_root_to_leaf_numbers_129/Solution.java
package sum_root_to_leaf_numbers_129;

import libraries.tree.binary_tree.TreeNode;

class Solution {
  public int sumNumbers(TreeNode root) {
    return sum(root, 0);
  }

  private static int sum(TreeNode root, int currentSum) {
    if (root == null) {
      return 0;
    }

    currentSum = currentSum * 10 + root.val;

    if (root.left == null && root.right == null) {
      return currentSum;
    }

    return sum(root.left, currentSum) + sum(root.right, currentSum);
  }
}

File: python/sum_root_to_leaf_numbers_129/Solution.py
from typing import Optional

from tree.binary_tree.TreeNode import TreeNode


class Solution:
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        return Solution.__sum(root, 0)

    @staticmethod
    def __sum(root: Optional[TreeNode], current: int) -> int:
        if not root:
            return 0

        current = current * 10 + root.val

        if not root.left and not root.right:
            return current

        return Solution.__sum(root.left, current) + Solution.__sum(root.right, current)